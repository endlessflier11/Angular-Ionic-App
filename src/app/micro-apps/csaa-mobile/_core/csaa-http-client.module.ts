import { NgModule } from '@angular/core';
import { AngularHttpAdapter } from './services/http/adapters/angular-http.adapter';
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { IonicHttpAdapter } from './services/http/adapters/ionic-http.adapter';
import { HTTP } from '@ionic-native/http/ngx';
import { UNI_HTTP_ADAPTER } from './services/http/uni-http.model';

@NgModule({
  imports: [HttpClientModule],
  providers: [
    HTTP,
    { provide: UNI_HTTP_ADAPTER, multi: true, useClass: AngularHttpAdapter, deps: [HttpClient] },
    {
      provide: UNI_HTTP_ADAPTER,
      multi: true,
      useClass: IonicHttpAdapter,
      deps: [HTTP],
    },
  ],
  exports: [HttpClientModule],
})
export class CsaaHttpClientModule {}
