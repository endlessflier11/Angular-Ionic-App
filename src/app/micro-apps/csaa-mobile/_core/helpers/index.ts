export * from './router.helpers';
export * from './misc.helpers';
export * from './async.helper';
export * from './error.helper';
export * from './string.helper';
