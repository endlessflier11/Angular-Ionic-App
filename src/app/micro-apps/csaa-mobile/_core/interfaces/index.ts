export * from './auth.interface';
export * from './router';
export * from './http.interface';
export * from './vehicle.interface';
export * from './ionic-lifecycle.interface';
export * from './analytics.interface';
export * from './policy.interface';
export * from './payment.interface';
export * from './config.interface';
export * from './ngxs-store.interface';
