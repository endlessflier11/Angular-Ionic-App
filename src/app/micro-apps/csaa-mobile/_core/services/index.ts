export * from './router/router.service';
export * from './config/config.service';
export * from './auth/auth.service';
export * from './sso/sso.service';
export * from './storage/storage.service';
export * from './http/http.service';
export * from './analytics.service';
export * from './policy.service';
export * from './payment.service';
export * from './global-state.service';
export * from './call.service';
export * from './metadata.service';
export * from './pdf-display.service';
export * from './webview/webview.service';
