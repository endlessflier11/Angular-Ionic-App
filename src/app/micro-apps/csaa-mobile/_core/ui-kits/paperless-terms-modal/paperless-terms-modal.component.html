<ion-header [ngClass]="['csaa', currentTheme]">
  <ion-toolbar no-border>
    <ion-buttons slot="start">
      <csaa-back-button (back)="dismissModal()"></csaa-back-button>
    </ion-buttons>
  </ion-toolbar>
</ion-header>
<ion-content [ngClass]="['csaa', currentTheme]">
  <csaa-skeleton-card *ngIf="loading"></csaa-skeleton-card>

  <ion-card class="csaa-card" *ngIf="!loading">
    <ng-container>
      <ion-card-header>
        <ion-card-title>Terms & Conditions</ion-card-title>
      </ion-card-header>
    </ng-container>
    <ion-card-content>
      <p>
        By tapping "Accept", you are accepting the
        <span class="csaa-link" (click)="openTerms()"
          >paperless preferences terms and conditions</span
        >
        and choosing to receive your policy documents and bills via email at {{ email }}. Policy
        documents include information about your policy, as well as privacy and other notices. Bill
        notifications include bills for insurance premiums and related fees. You may opt out of
        paperless preferences at any time, or update your email address by:
      </p>
      <br />
      <ul>
        <li>Editing your Paperless Settings in MyPolicy</li>
        <li>Calling Insurance Service at <a href="tel:+18883401817">888.340.1817</a></li>
        <li>Visiting your local AAA branch</li>
      </ul>
      <ion-button expand="block" class="ion-margin" (click)="acceptClickHandler()"
        >Accept
      </ion-button>
    </ion-card-content>
  </ion-card>
</ion-content>
