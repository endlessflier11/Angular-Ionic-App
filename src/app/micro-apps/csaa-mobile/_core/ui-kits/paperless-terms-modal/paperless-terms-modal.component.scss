ion-content {
  --background: #fff;
}

ion-card {
  box-shadow: none;
}
.csaa-link {
  text-decoration: underline;
}
