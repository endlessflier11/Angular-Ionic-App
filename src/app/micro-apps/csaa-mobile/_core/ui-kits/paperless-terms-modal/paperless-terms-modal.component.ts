import { Component, Input, OnInit } from '@angular/core';
import { noop } from 'rxjs';
import { Store } from '@ngxs/store';
import { ModalController } from '@ionic/angular';
import { ConfigState } from '../../../_core/store/states/config.state';
import { finalize, map, switchMap } from 'rxjs/operators';
import { CustomerAction, PolicyAction, UserSettingAction } from '../../../_core/store/actions';
import { interactWithLoader } from '../../../_core/operators';
import {
  AppEndpointsEnum,
  Category,
  EventName,
  EventType,
  Policy,
} from '../../../_core/interfaces';
import { AnalyticsService, WebviewService } from '../../../_core/services';
import { UserSettingsState } from '../../store/states/user-setting.state';
import { MetadataState } from '../../store/states/metadata.state';
import { ConfigSelector } from '../../store/selectors';

@Component({
  selector: 'csaa-paperless-terms-modal',
  templateUrl: './paperless-terms-modal.component.html',
  styleUrls: ['./paperless-terms-modal.component.scss'],
})
export class PaperlessTermsModalComponent implements OnInit {
  currentTheme: string;
  email = '';
  loading = false;
  @Input() policy: Policy;

  constructor(
    private readonly store: Store,
    private readonly modalCtrl: ModalController,
    private readonly analyticsService: AnalyticsService,
    private readonly webviewService: WebviewService
  ) {}

  ngOnInit() {
    this.currentTheme = this.store.selectSnapshot(ConfigState.theme);
    this.loading = true;
    this.store
      .dispatch(new CustomerAction.LoadCustomer())
      .pipe(
        switchMap(() => this.store.dispatch(new PolicyAction.LoadPolicies())),
        switchMap(() =>
          this.store.dispatch(new UserSettingAction.LoadPolicyPaperlessPreferences())
        ),
        map(() => {
          const paperless = this.store.selectSnapshot(UserSettingsState.paperlessPreferences);
          // 1) If policy is provided, try to get the email associated with the policy
          let email = paperless?.find((p) => p.policyNumber === this.policy?.number)?.email?.email;
          // 2) Or else, try to get any email from paperless preferences
          email = email || paperless?.reduce((e, p) => e || p?.email?.email, '');
          // 3) As a fallback, get the email from user login
          return email || this.store.selectSnapshot(MetadataState)?.email;
        }),
        finalize(() => (this.loading = false))
      )
      .subscribe((e) => (this.email = e));
  }

  dismissModal(withSuccess?: boolean) {
    this.modalCtrl.dismiss(withSuccess).then(noop);
  }

  acceptClickHandler(): void {
    this.analyticsService.trackEvent(EventName.TERMS_AND_CONDITIONS_ANSWERED, Category.documents, {
      event_type: EventType.OPTION_SELECTED,
      selection: 'Accept',
    });
    this.store
      .dispatch(new CustomerAction.AcceptPaperlessTerms())
      .pipe(interactWithLoader())
      .subscribe(() => this.dismissModal(true));
  }

  async openTerms() {
    this.analyticsService.trackEvent(EventName.TERMS_AND_CONDITIONS_CLICKED, Category.documents, {
      event_type: EventType.FILE_DOWNLOADED,
      file: 'Paperless Terms and Conditions',
    });
    // TODO: Adding these cookies is not preventing the user to be redirected
    // const { zipCode, clubCode } = this.store.selectSnapshot(ConfigState);
    // const cookies: CookieJar[] = [
    //   {
    //     domain: '.aaa.com',
    //     name: 'AAAClubCode',
    //     value: `${clubCode}`,
    //   },
    //   {
    //     domain: '.aaa.com',
    //     name: 'AAAZIPCode',
    //     value: `${zipCode}`,
    //   },
    // ];
    const {
      appEndpoints: { endpoints },
    } = this.store.selectSnapshot(ConfigSelector.state);
    const url = endpoints[AppEndpointsEnum[AppEndpointsEnum.paperlessTermsOfUseUrl]];
    await this.webviewService.openWithCookies(url, []);
  }
}
