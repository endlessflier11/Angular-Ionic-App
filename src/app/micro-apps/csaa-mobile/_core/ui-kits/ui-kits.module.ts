import { NgModule } from '@angular/core';
import { CommonModule, DatePipe } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { IonicModule } from '@ionic/angular';
import { ThemeIconComponent } from './theme-icon/theme-icon.component';
import { WorkingHoursComponent } from './working-hours/working-hours.component';
import { MyPolicyCardComponent } from './my-policy-card/my-policy-card.component';
import { SkeletonCardComponent } from './skeleton-card/skeleton-card.component';
import { BackButtonComponent } from './back-button/back-button.component';
import { PhonePipe } from './pipes/phone-number.pipe';
import { CsaaAutoIdCardComponent } from './csaa-auto-id-card/csaa-auto-id-card.component';
import { CsaaLoadingComponent } from './csaa-loading/csaa-loading.component';
import { ContentRendererComponent } from './content-renderer/content-renderer.component';
import { NationalCustomAlertModalComponent } from './national-custom-alert/national-custom-alert-modal';
import { SafeDataPipe } from './pipes/safe-data/safe-data.pipe';
import { DocumentActionModalComponent } from './document-action-modal/document-action-modal.component';
import { CsaaPolicySubtitlesComponent } from './csaa-policy-subtitles/csaa-policy-subtitles.component';
import { PaperlessTermsComponent } from './paperless-terms/paperless-terms.component';
import { PaperlessTermsModalComponent } from './paperless-terms-modal/paperless-terms-modal.component';
import { AppleWalletButtonComponent } from './apple-wallet-button/apple-wallet-button.component';

const UI_COMPONENTS = [
  ThemeIconComponent,
  WorkingHoursComponent,
  MyPolicyCardComponent,
  SkeletonCardComponent,
  BackButtonComponent,
  CsaaAutoIdCardComponent,
  CsaaLoadingComponent,
  ContentRendererComponent,
  NationalCustomAlertModalComponent,
  DocumentActionModalComponent,
  CsaaPolicySubtitlesComponent,
  PaperlessTermsComponent,
  PaperlessTermsModalComponent,
  AppleWalletButtonComponent,
];

const UI_PIPES = [PhonePipe, SafeDataPipe];

@NgModule({
  imports: [CommonModule, ReactiveFormsModule, IonicModule],
  declarations: [...UI_COMPONENTS, ...UI_PIPES],
  exports: [...UI_COMPONENTS, ...UI_PIPES],
  providers: [DatePipe],
})
export class UiKitsModule {}
