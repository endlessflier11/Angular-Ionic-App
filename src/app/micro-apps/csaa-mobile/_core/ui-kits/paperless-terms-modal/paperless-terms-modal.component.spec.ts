import { ComponentFixture, TestBed } from '@angular/core/testing';
import { IonicModule } from '@ionic/angular';

import { PaperlessTermsModalComponent } from './paperless-terms-modal.component';
import { UiKitsModule } from '../../../_core/ui-kits/ui-kits.module';
import { PageTestingModule, StoreTestBuilder } from '@app/testing';
import { Store } from '@ngxs/store';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { UniHttpAdapterTarget } from '../../../_core/interfaces';
import { HttpService } from '../../services';
import { CsaaHttpClientModule } from '../../csaa-http-client.module';
import { CsaaCoreModule } from '../../../csaa-core/csaa-core.module';

describe('PaperlessTermsModalComponent', () => {
  let component: PaperlessTermsModalComponent;
  let fixture: ComponentFixture<PaperlessTermsModalComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        IonicModule.forRoot(),
        CsaaCoreModule,
        UiKitsModule,
        HttpClientTestingModule,
        CsaaHttpClientModule,
        PageTestingModule.withConfig({
          providesStorage: true,
          providesAnalytics: true,
          providesPlatform: true,
          providesWebviewService: true,
          providesRouter: true,
        }),
      ],
    }).compileComponents();

    const store = TestBed.inject(Store);
    const httpService = TestBed.inject(HttpService);
    (httpService as any).currentAdapterTarget = UniHttpAdapterTarget.NATIVE;
    StoreTestBuilder.withDefaultMocks().resetStateOf(store);

    fixture = TestBed.createComponent(PaperlessTermsModalComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
