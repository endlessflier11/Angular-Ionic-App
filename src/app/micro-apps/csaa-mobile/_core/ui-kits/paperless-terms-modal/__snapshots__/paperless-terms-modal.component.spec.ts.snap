// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaperlessTermsModalComponent should match snapshot 1`] = `
<csaa-paperless-terms-modal
  __ngContext__={[Function LRootView]}
  analyticsService={[Function AnalyticsMockService]}
  currentTheme=""
  email={[Function String]}
  loading="false"
  modalCtrl={[Function ModalController]}
  store={[Function Store]}
  webviewService={[Function Function]}
>
  <ion-header
    class="csaa"
  >
    <ion-toolbar
      no-border=""
    >
      <ion-buttons
        slot="start"
      >
        <csaa-back-button>
          <ion-button>
            <ion-icon
              name="chevron-back-outline"
              slot="start"
            />
             Back
          </ion-button>
        </csaa-back-button>
      </ion-buttons>
    </ion-toolbar>
  </ion-header><ion-content
    class="csaa"
  >
    <ion-card
      class="csaa-card"
    >
      <ion-card-header>
        <ion-card-title>
          Terms & Conditions
        </ion-card-title>
      </ion-card-header>
      <ion-card-content>
        <p>
           By tapping "Accept", you are accepting the 
          <span
            class="csaa-link"
          >
            paperless preferences terms and conditions
          </span>
           and choosing to receive your policy documents and bills via email at A151@TESTUSER.EXAMPLE.COM. Policy documents include information about your policy, as well as privacy and other notices. Bill notifications include bills for insurance premiums and related fees. You may opt out of paperless preferences at any time, or update your email address by: 
        </p>
        <br />
        <ul>
          <li>
            Editing your Paperless Settings in MyPolicy
          </li>
          <li>
            Calling Insurance Service at 
            <a
              href="tel:+18883401817"
            >
              888.340.1817
            </a>
          </li>
          <li>
            Visiting your local AAA branch
          </li>
        </ul>
        <ion-button
          class="ion-margin"
          expand="block"
        >
          Accept 
        </ion-button>
      </ion-card-content>
    </ion-card>
  </ion-content>
</csaa-paperless-terms-modal>
`;
